java/server/src/org/openqa/grid/common/SeleniumProtocol.java

```diff
@@ -17,8 +17,50 @@
 
 package org.openqa.grid.common;
 
+import static org.openqa.grid.common.RegistrationRequest.PATH;
+import static org.openqa.grid.common.RegistrationRequest.SELENIUM_PROTOCOL;
+
+import org.openqa.grid.common.exception.GridException;
+import org.openqa.selenium.remote.DesiredCapabilities;
+
+import java.util.Arrays;
+
 public enum SeleniumProtocol {
-  Selenium, WebDriver;
+  Selenium("/selenium-server/driver"),
+  WebDriver("/wd/hub");
+  private String path;
+
+  SeleniumProtocol(String path) {
+    this.path = path;
+  }
+
+  public String getPath() {
+    return path;
+  }
+
+  public static SeleniumProtocol getProtocol(DesiredCapabilities capability) {
+    String type = (String) capability.getCapability(SELENIUM_PROTOCOL);
+    if (type == null || type.trim().isEmpty()) {
+      SeleniumProtocol protocol = WebDriver;
+      protocol.computePath(capability);
+      return protocol;
+    }
+    try {
+      SeleniumProtocol protocol = SeleniumProtocol.valueOf(type);
+      protocol.computePath(capability);
+      return protocol;
+    } catch (IllegalArgumentException e) {
+      throw new GridException(type + " isn't a valid protocol type for grid. Valid values :[" +
+                              Arrays.toString(values()) + "]", e);
+    }
+  }
+
+  private void computePath(DesiredCapabilities capability) {
+    String localPath = (String) capability.getCapability(PATH);
+    if (localPath != null) {
+      this.path = localPath;
```

Comment:
i think this is pretty dangerous, you're changing the enum's value of path. enums are static, so you're changing the path for *every* instance of the protocol type.

I'm imagining a hub that has multiple nodes with different paths.... if one changes the path, then all subsequent will have that new path, rather than the default /wd/hub path that they might assume.